src/audio/smart_amp/smart_amp_maxim_dsm.c

```diff
@@ -0,0 +1,459 @@
+	buf.buf32 = (int32_t *)hspk->buf.fb.buf;
+
+	if (*w_ptr + nsamples <= DSM_FB_BUF_DB_SZ) {
+		if (is_16bit)
+			memcpy_s(&buf.buf16[*w_ptr], nsamples * szsample,
+				 in, nsamples * szsample);
+		else
+			memcpy_s(&buf.buf32[*w_ptr], nsamples * szsample,
+				 in, nsamples * szsample);
+		*w_ptr += nsamples;
+	} else {
+		comp_err(dev, "[DSM] Feedback buffer overflow. w_ptr : %d",
+			 *w_ptr);
+		return;
+	}
+	if (*w_ptr >= DSM_FB_BUF_SZ) {
+		/* Do FB processing */
+		for (x = 0; x < DSM_FRM_SZ; x++) {
+			/* Buffer ordering for DSM : IVIV... -> VV... II...*/
+			if (is_16bit) {
+				v[x] = buf.buf16[4 * x + 1];
+				i[x] = buf.buf16[4 * x];
+				v[x + DSM_FRM_SZ] = buf.buf16[4 * x + 3];
+				i[x + DSM_FRM_SZ] = buf.buf16[4 * x + 2];
+			} else {
+				v[x] = (buf.buf32[4 * x + 1] >> 16);
+				i[x] = (buf.buf32[4 * x] >> 16);
+				v[x + DSM_FRM_SZ] =
+					(buf.buf32[4 * x + 3] >> 16);
+				i[x + DSM_FRM_SZ] =
+					(buf.buf32[4 * x + 2] >> 16);
+			}
+		}
+
+		remain = (*w_ptr - DSM_FB_BUF_SZ);
+		if (remain) {
+			if (is_16bit)
+				_memmove(&buf.buf16[0],
+					 &buf.buf16[DSM_FB_BUF_SZ],
+					 remain * szsample);
+			else
+				_memmove(&buf.buf32[0],
+					 &buf.buf32[DSM_FB_BUF_SZ],
+					 remain * szsample);
+		}
+		*w_ptr -= DSM_FB_BUF_SZ;
+
+		/* Do DSM feedback processing here */
+	}
+}
+
+int smart_amp_flush(struct smart_amp_mod_struct_t *hspk, struct comp_dev *dev)
+{
+	memset(hspk->buf.frame_in, 0,
+	       SMART_AMP_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(hspk->buf.frame_out, 0,
+	       SMART_AMP_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(hspk->buf.frame_iv, 0,
+	       SMART_AMP_FB_BUF_DB_SZ * sizeof(int32_t));
+
+	memset(hspk->buf.input, 0, DSM_FF_BUF_SZ * sizeof(int16_t));
+	memset(hspk->buf.output, 0, DSM_FF_BUF_SZ * sizeof(int16_t));
+	memset(hspk->buf.voltage, 0, DSM_FF_BUF_SZ * sizeof(int16_t));
+	memset(hspk->buf.current, 0, DSM_FF_BUF_SZ * sizeof(int16_t));
+
+	memset(hspk->buf.ff.buf, 0, DSM_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(hspk->buf.ff_out.buf, 0, DSM_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(hspk->buf.fb.buf, 0, DSM_FB_BUF_DB_SZ * sizeof(int32_t));
+
+	hspk->buf.ff.avail = DSM_FF_BUF_SZ;
+	hspk->buf.ff_out.avail = 0;
+	hspk->buf.fb.avail = 0;
+
+	comp_dbg(dev, "[DSM] Reset (handle:%p)", (uintptr_t)hspk);
+
+	return 0;
+}
+
+int smart_amp_init(struct smart_amp_mod_struct_t *hspk, struct comp_dev *dev)
+{
+	return maxim_dsm_init(hspk, dev);
+}
+
+static int smart_amp_get_buffer(int32_t *buf, uint32_t frames,
+				struct audio_stream *stream,
+				int8_t *chan_map, uint32_t num_ch)
+{
+	int x, y;
+	uint32_t in_frag = 0;
+	union smart_amp_buf input, output;
+	int index;
+
+	input.buf16 = (int16_t *)stream->r_ptr;
+	input.buf32 = (int32_t *)stream->r_ptr;
+	output.buf16 = (int16_t *)buf;
+	output.buf32 = (int32_t *)buf;
+
+	switch (stream->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		for (x = 0 ; x < frames ; x++) {
+			for (y = 0 ; y < num_ch; y++) {
+				if (chan_map[y] == -1)
+					continue;
```

Comment:
If x is chan_map then this would mean you need to check this less often in the nested loop. i.e. you could check it in the outer loop.

Reply:
Each frame has 'num_ch' samples. audio_stream_write_frag_s16 function gets the data in this format.
If I bring chan_map[y] to the outer loop, data format will be broken.